### PR TITLE
Provena Release v2.4

### PR DESCRIPTION
**Pull request by PeterBaker0**
## Provena V2.4 Release

Includes the following changes

### Minor

- RRAPIS-1580 (Minor): Moving model version annotation from workflow template to model run (#68)
    Moves the "software_version" field from the model run workflow template to the model run itself, and renames it "model_version".

    This change will allow modellers to annotate model versions more naturally. For example, it is unlikely that every time a model was changed (say in a git repo), that a new workflow template would be desired. Instead, the modeller can annotate the git hash or other version identifier for each model run, reusing the same template. The model run workflow template retains it's own versioning capabilities, where required.

    The migration finds all model run workflow templates, and moves the version annotation to all model runs which reference it.

    I have updated the docs where appropriate to reference this new field.

- RRAPIS-1583 (Minor): Data store API S3 path code refactor (#69)

    improves authorisation s3 path code due to identified poor code quality.

- RRAPIS-1575 (Minor): Improved graph restore functionality utilising job system and expanding to all provenance types  (#66)

    Adds a restore graph functionality which utilises the job system efficiently to restore provenance only for existing registry items.

    Locking moto library

    Adding a linked/unlinked dataset + summary stats functionality.

- RRAPIS-1581 (Minor): allow storage of files for externally reposited datasets (#67)

### Move MRWT.software_version to model_run.model_version

The migration finds all model run workflow templates, and moves the version annotation to all model runs which reference it.

```bash
# To test and validate
./migrate f1580-migration <STAGE>
# To apply
./migrate f1580-migration <STAGE> --apply
```

### CODE REVIEW

File: data-store-api/routes/registry/credentials.py [hunk 1]
```diff
@@ -31,6 +31,8 @@ def allows_credential_generation(dataset: ItemDataset) -> bool:
 
     Ensures that the dataset is reposited before enabling credential generation.
 
+    NOTE: RRAPIS-1581 changed this behaviour to enable storage access even if externally reposited.
```
Comment (Rosspet) [inline data-store-api/routes/registry/credentials.py:34]: ⚠️ ⚠️ ⚠️ ⚠️ PLS REMOVE ⚠️ ⚠️ ⚠️ ⚠️  NO RRAP REFERENCE ⚠️ ⚠️ ⚠️ ⚠️

File: data-store-api/routes/registry/credentials.py [hunk 2]
```diff
@@ -41,7 +43,11 @@ def allows_credential_generation(dataset: ItemDataset) -> bool:
     bool
         True iff the dataset should enable credential generation
     """
-    return dataset.collection_format.dataset_info.access_info.reposited
+    # NOTE JIRA RRAPIS-1581 re-enabled data storage even for externally reposited datasets
```
Comment (Rosspet) [inline data-store-api/routes/registry/credentials.py:46]: 🅱️

File: data-store-api/routes/registry/credentials.py (hunk 1 above)
Comment (PeterBaker0) [reply data-store-api/routes/registry/credentials.py:34]: Think we'll be okay!

**Review by Rosspet — APPROVED** [review]
lgtm

**Pull request merged**